tools/build_defs/repo/utils.bzl

```diff
@@ -91,6 +91,30 @@ def _download_patch(ctx, patch_url, integrity, auth):
     )
     return patch_path
 
+def remote_files(ctx, auth = None):
+    """Utility function for downloading remote files.
+
+    This rule is intended to be used in the implementation function of
+    a repository rule. It assumes the parameters `remote_file_urls` and
+    `remote_file_integrity` to be present in `ctx.attr`.
+
+    Args:
+      ctx: The repository context of the repository rule calling this utility
+        function.
+      auth: An optional dict specifying authentication information for some of the URLs.
+    """
+    for path, remote_file_urls in ctx.attr.remote_file_urls.items():
+        # integrity is optional but really recommended
+        integrity = ctx.attr.remote_file_integrity.get(path)
+
+        ctx.download(
```

Comment:
Appreciate advice here on how we might want to check for relative (..) path invocations.
Are there utility functions to employ?

Reply:
I'm not aware of any since all the other logic dealing with this (e.g. decompression or patching) isn't in Starlark.

I could see this working:
```starlark
repo_dir = ctx.path(".")
resolved_repo_dir = repo_dir.realpath
path = repo_dir.get_child(relative_path)
resolved_path = path.realpath
if not str(resolved_path).starts_with(str(resolved_repo_dir) + "/"):
  fail("...")
```

Reply:
I had to add this validation after the files were downloaded which is a bit silly on account of https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/vfs/Path.java#L530

Calling `realpath` throws exception AFAICT if the path does not exist.